File: server/chatbot.js
// chatbot.js
require("dotenv").config();
const express = require("express");
const Groq = require("groq-sdk");
const cors = require("cors");
const { Pinecone } = require("@pinecone-database/pinecone");
const { scrapeWebsite } = require("./scraper");

// Create Express app
const app = express();

// Middleware
app.use(
  cors({
    origin: "*", 
    methods: ["POST", "GET"],
    credentials: true,
  })
);
app.use(express.json());

// Groq configuration
const GROQ_API_KEY = process.env.GROQ_API_KEY;
const PINECONE_API_KEY = process.env.PINECONE_API_KEY;
const LLM_MODEL = "openai/gpt-oss-20b";
const PINECONE_INDEX_NAME = "chatbot";

if (!GROQ_API_KEY) {
  console.error("ERROR: GROQ_API_KEY environment variable is not set!");
  console.error("Please set it using: export GROQ_API_KEY='your-api-key-here'");
  process.exit(1);
}

if (!PINECONE_API_KEY) {
  console.error("ERROR: PINECONE_API_KEY environment variable is not set!");
  console.error("Please set it using: export PINECONE_API_KEY='your-api-key-here'");
  process.exit(1);
}

// Initialize Groq client
const groq = new Groq({
  apiKey: GROQ_API_KEY,
});

// Initialize Pinecone client
const pinecone = new Pinecone({
  apiKey: PINECONE_API_KEY,
});

// Track initialization
let isInitialized = false;
let pineconeIndex = null;

/**
 * Generate embeddings using Pinecone's inference API
 */
async function generateEmbedding(text) {
  try {
    const embeddings = await pinecone.inference.embed(
      "llama-text-embed-v2",
      [text],
      { inputType: "passage" }
    );
    
    return embeddings[0].values;
  } catch (error) {
    console.error("Error generating embedding:", error.message);
    throw error;
  }
}

/**
 * Initialize the chatbot system
 */
async function initializeChatbot() {
  try {
    // Get the Pinecone index
    pineconeIndex = pinecone.index(PINECONE_INDEX_NAME);
    
    isInitialized = true;
    console.log("Chatbot initialized successfully");
    console.log(`Connected to Pinecone index: ${PINECONE_INDEX_NAME}`);
  } catch (error) {
    console.error("Failed to initialize chatbot:", error.message);
    throw error;
  }
}

/**
 * Store chunks in Pinecone
 */
async function storeChunks(chunks, metadata) {
  try {
    const vectors = [];
    
    for (let i = 0; i < chunks.length; i++) {
      const chunk = chunks[i];
      const embedding = await generateEmbedding(chunk);
      
      vectors.push({
        id: `chunk_${Date.now()}_${i}`,
        values: embedding,
        metadata: {
          text: chunk,
          url: metadata.url,
          title: metadata.title,
          timestamp: new Date().toISOString(),
        },
      });
      
      // Log progress
      if ((i + 1) % 10 === 0) {
        console.log(`Generated embeddings for ${i + 1}/${chunks.length} chunks`);
      }
    }
    
    // Upsert vectors to Pinecone in batches
    const batchSize = 100;
    for (let i = 0; i < vectors.length; i += batchSize) {
      const batch = vectors.slice(i, i + batchSize);
      await pineconeIndex.upsert(batch);
      console.log(`Upserted batch ${Math.floor(i / batchSize) + 1}/${Math.ceil(vectors.length / batchSize)}`);
    }
    
    return vectors.length;
  } catch (error) {
    console.error("Error storing chunks:", error.message);
    throw error;
  }
}

/**
 * Search for similar chunks in Pinecone
 */
async function searchSimilar(query, topK = 5) {
  try {
    // Generate embedding for the query
    const queryEmbedding = await generateEmbedding(query);
    
    // Search in Pinecone
    const results = await pineconeIndex.query({
      vector: queryEmbedding,
      topK: topK,
      includeMetadata: true,
    });
    
    // Format results
    return results.matches.map(match => ({
      text: match.metadata.text,
      url: match.metadata.url,
      title: match.metadata.title,
      score: match.score,
    }));
  } catch (error) {
    console.error("Error searching similar chunks:", error.message);
    throw error;
  }
}

/**
 * Get count of vectors in Pinecone
 */
async function getCount() {
  try {
    const stats = await pineconeIndex.describeIndexStats();
    return stats.totalRecordCount || 0;
  } catch (error) {
    console.error("Error getting count:", error.message);
    return 0;
  }
}

/**
 * Clear all vectors from Pinecone
 */
async function clearAll() {
  try {
    await pineconeIndex.deleteAll();
    console.log("All vectors cleared from Pinecone");
  } catch (error) {
    console.error("Error clearing vectors:", error.message);
    throw error;
  }
}

/**
 * Generate response using Groq LLM
 */
async function generateLLMResponse(prompt, context) {
  try {
    const systemPrompt = `You are a helpful assistant that answers questions based solely on the provided context from a website.

IMPORTANT RULES:
1. Only answer based on the context provided below
2. If the answer cannot be found in the context, say: "I could not find that information on the website."
3. Be concise and accurate
4. Do not make up information
5. If relevant, you can mention which part of the website the information comes from

Context:
${context}`;

    const completion = await groq.chat.completions.create({
      messages: [
        {
          role: "system",
          content: systemPrompt,
        },
        {
          role: "user",
          content: prompt,
        },
      ],
      model: LLM_MODEL,
      temperature: 0.3,
      max_tokens: 1024,
      top_p: 0.9,
    });

    return completion.choices[0]?.message?.content || "I apologize, but I couldn't generate a response.";
  } catch (error) {
    console.error("Error generating LLM response:", error.message);
    if (error.status === 401) {
      throw new Error("Invalid Groq API key. Please check your GROQ_API_KEY environment variable.");
    }
    if (error.status === 429) {
      throw new Error("Rate limit exceeded. Please try again later.");
    }
    throw error;
  }
}

/**
 * Process user query with RAG pipeline
 */
async function processQuery(userQuery) {
  try {
    // Step 1: Retrieve relevant chunks
    const relevantChunks = await searchSimilar(userQuery, 5);

    if (relevantChunks.length === 0) {
      return {
        answer:
          "I could not find that information on the website. The knowledge base might be empty or your question might not be covered.",
        sources: [],
      };
    }

    // Step 2: Prepare context from retrieved chunks
    const context = relevantChunks
      .map((chunk, idx) => `[Source ${idx + 1}]: ${chunk.text}`)
      .join("\n\n");

    console.log(`Retrieved ${relevantChunks.length} relevant chunks`);

    // Step 3: Generate answer using LLM
    const answer = await generateLLMResponse(userQuery, context);

    // Step 4: Prepare sources
    const sources = relevantChunks.map((chunk, idx) => ({
      index: idx + 1,
      text: chunk.text.substring(0, 200) + "...",
      url: chunk.url,
      title: chunk.title,
      relevanceScore: chunk.score,
    }));

    return {
      answer,
      sources,
    };
  } catch (error) {
    console.error("Error processing query:", error.message);
    throw error;
  }
}

/**
 * POST /getMsg - Main chatbot endpoint
 */
app.post("/getMsg", async (req, res) => {
  try {
    const { content, role } = req.body;

    if (!content || role !== "user") {
      return res.status(400).json({
        error: 'Invalid request. Expected { content: string, role: "user" }',
      });
    }

    console.log(`Received query: ${content}`);

    // Process the query through RAG pipeline
    const result = await processQuery(content);

    res.json({
      role: "assistant",
      content: result.answer,
      sources: result.sources,
      timestamp: new Date().toISOString(),
    });
  } catch (error) {
    console.error("Error in /getMsg:", error.message);
    res.status(500).json({
      error: "An error occurred while processing your request",
      details: error.message,
    });
  }
});

/**
 * POST /scrape - Scrape and index a website
 */
app.post("/scrape", async (req, res) => {
  try {
    const url = "https://agiteks.com";
    const { isDynamic = false } = req.body;

    if (!url) {
      return res.status(400).json({
        error: "URL is required",
      });
    }

    console.log(`Starting scrape for: ${url}`);

    // Scrape the website
    const scrapedData = await scrapeWebsite(url, isDynamic);

    // Store chunks in Pinecone
    const count = await storeChunks(scrapedData.chunks, {
      url: scrapedData.url,
      title: scrapedData.title,
    });

    res.json({
      message: "Website scraped and indexed successfully",
      url: scrapedData.url,
      title: scrapedData.title,
      chunksStored: count,
      timestamp: scrapedData.scrapedAt,
    });
  } catch (error) {
    console.error("Error in /scrape:", error.message);
    res.status(500).json({
      error: "Failed to scrape website",
      details: error.message,
    });
  }
});

/**
 * GET /status - Check chatbot status
 */
app.get("/status", async (req, res) => {
  try {
    const count = await getCount();

    res.json({
      status: "running",
      initialized: isInitialized,
      chunksInDatabase: count,
      vectorDB: "Pinecone",
      indexName: PINECONE_INDEX_NAME,
      embeddingModel: "llama-text-embed-v2",
      llmModel: LLM_MODEL,
      apiKeysConfigured: {
        groq: !!GROQ_API_KEY,
        pinecone: !!PINECONE_API_KEY,
      },
    });
  } catch (error) {
    res.status(500).json({
      error: "Error getting status",
      details: error.message,
    });
  }
});

/**
 * DELETE /clear - Clear all data from database
 */
app.delete("/clear", async (req, res) => {
  try {
    await clearAll();

    res.json({
      message: "Pinecone index cleared successfully",
    });
  } catch (error) {
    res.status(500).json({
      error: "Error clearing database",
      details: error.message,
    });
  }
});

/**
 * Start the server
 */
const PORT = process.env.PORT || 3000;

async function startServer() {
  try {
    // Initialize the chatbot system
    await initializeChatbot();

    // Start Express server
    app.listen(PORT, () => {
      console.log(
        `\n🤖 RAG Chatbot Server (Groq + Pinecone) running on http://localhost:${PORT}`
      );
      console.log(`\nAvailable endpoints:`);
      console.log(`  POST /getMsg    - Send a message to the chatbot`);
      console.log(`  POST /scrape    - Scrape and index a website`);
      console.log(`  GET  /status    - Check system status`);
      console.log(`  DELETE /clear   - Clear all data`);
      console.log(`\nUsing:`);
      console.log(`  LLM: ${LLM_MODEL} (Groq)`);
      console.log(`  Vector DB: Pinecone (${PINECONE_INDEX_NAME})`);
      console.log(`  Embeddings: llama-text-embed-v2 (Pinecone)\n`);
    });
  } catch (error) {
    console.error("Failed to start server:", error.message);
    process.exit(1);
  }
}

// Start the server
startServer();

module.exports = app;

File: server/chatbot.py
"""
RAG Chatbot using LangChain, Groq, and FAISS
Scrapes agiteks.com and answers queries based on that data

Prerequisites:
pip install langchain langchain-groq langchain-community python-dotenv
pip install faiss-cpu beautifulsoup4 requests
pip install langchain-huggingface sentence-transformers
"""

import os
import pickle
import requests
from bs4 import BeautifulSoup
from urllib.parse import urljoin, urlparse
from dotenv import load_dotenv

from langchain_groq import ChatGroq
from langchain.schema import HumanMessage, AIMessage, SystemMessage
from langchain.text_splitter import RecursiveCharacterTextSplitter
from langchain_huggingface import HuggingFaceEmbeddings
from langchain_community.vectorstores import FAISS

load_dotenv()

# Configuration
WEBSITE_URL = "https://agiteks.com"
FAISS_INDEX_PATH = "faiss_index"
MAX_PAGES_TO_SCRAPE = 20

class WebsiteScraper:
    """Scrapes website content"""
    
    def __init__(self, base_url, max_pages=20):
        self.base_url = base_url
        self.max_pages = max_pages
        self.visited_urls = set()
        self.domain = urlparse(base_url).netloc
    
    def is_valid_url(self, url):
        """Check if URL belongs to the same domain"""
        parsed = urlparse(url)
        return parsed.netloc == self.domain
    
    def scrape_page(self, url):
        """Scrape content from a single page"""
        try:
            headers = {
                'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36'
            }
            response = requests.get(url, timeout=10, headers=headers)
            response.raise_for_status()
            soup = BeautifulSoup(response.content, 'html.parser')
            
            # Remove script and style elements
            for script in soup(["script", "style", "nav", "footer", "header"]):
                script.decompose()
            
            # Get text content
            text = soup.get_text(separator=' ', strip=True)
            
            # Get all links for crawling
            links = []
            for link in soup.find_all('a', href=True):
                absolute_url = urljoin(url, link['href'])
                if self.is_valid_url(absolute_url) and absolute_url not in self.visited_urls:
                    links.append(absolute_url)
            
            return text, links
        
        except Exception as e:
            print(f"Error scraping {url}: {e}")
            return None, []
    
    def scrape_website(self):
        """Scrape multiple pages from the website"""
        to_visit = [self.base_url]
        documents = []
        
        print(f"Starting to scrape {self.base_url}...")
        
        while to_visit and len(self.visited_urls) < self.max_pages:
            url = to_visit.pop(0)
            
            if url in self.visited_urls:
                continue
            
            print(f"Scraping: {url} ({len(self.visited_urls) + 1}/{self.max_pages})")
            self.visited_urls.add(url)
            
            text, links = self.scrape_page(url)
            
            if text and len(text.strip()) > 100:
                documents.append({
                    'content': text,
                    'source': url
                })
            
            # Add new links to visit
            to_visit.extend([link for link in links if link not in self.visited_urls])
        
        print(f"\nScraped {len(documents)} pages successfully!")
        return documents

class RAGSystem:
    """RAG system with FAISS vector database"""
    
    def __init__(self):
        print("Initializing embeddings model...")
        self.embeddings = HuggingFaceEmbeddings(
            model_name="sentence-transformers/all-MiniLM-L6-v2"
        )
        self.vectorstore = None
        print("Embeddings model loaded!")
    
    def process_and_store_documents(self, documents):
        """Split documents and store in FAISS"""
        print("\nProcessing documents...")
        
        # Split documents into chunks
        text_splitter = RecursiveCharacterTextSplitter(
            chunk_size=1000,
            chunk_overlap=200,
            length_function=len
        )
        
        texts = []
        metadatas = []
        
        for doc in documents:
            chunks = text_splitter.split_text(doc['content'])
            texts.extend(chunks)
            metadatas.extend([{'source': doc['source']} for _ in chunks])
        
        print(f"Created {len(texts)} text chunks")
        
        # Create FAISS vectorstore
        print("Creating embeddings and storing in FAISS...")
        self.vectorstore = FAISS.from_texts(
            texts=texts,
            embedding=self.embeddings,
            metadatas=metadatas
        )
        
        # Save to disk
        print(f"Saving FAISS index to {FAISS_INDEX_PATH}...")
        self.vectorstore.save_local(FAISS_INDEX_PATH)
        
        print("Documents stored successfully!")
    
    def load_existing_vectorstore(self):
        """Load existing FAISS vectorstore"""
        if not os.path.exists(FAISS_INDEX_PATH):
            raise FileNotFoundError(f"FAISS index not found at {FAISS_INDEX_PATH}")
        
        print(f"Loading existing FAISS index from {FAISS_INDEX_PATH}...")
        self.vectorstore = FAISS.load_local(
            FAISS_INDEX_PATH,
            self.embeddings,
            allow_dangerous_deserialization=True
        )
        print("FAISS index loaded successfully!")
    
    def retrieve_context(self, query, k=3):
        """Retrieve relevant context for a query"""
        if not self.vectorstore:
            return ""
        
        docs = self.vectorstore.similarity_search(query, k=k)
        
        # Format context with sources
        context_parts = []
        for i, doc in enumerate(docs, 1):
            source = doc.metadata.get('source', 'Unknown')
            context_parts.append(f"[Source {i}: {source}]\n{doc.page_content}")
        
        context = "\n\n".join(context_parts)
        return context

def chatbot_with_rag():
    """Main chatbot function with RAG"""
    print("=" * 50)
    print("RAG Chatbot (Groq + LangChain + FAISS)")
    print("=" * 50)
    
    # Initialize LLM
    llm = ChatGroq(
        model="openai/gpt-oss-20b", 
        temperature=0.7,
        max_tokens=1024
    )
    
    # Initialize RAG system
    rag = RAGSystem()
    
    # Check if we need to scrape and index
    print("\nChecking for existing FAISS index...")
    
    if os.path.exists(FAISS_INDEX_PATH):
        choice = input("FAISS index found. Do you want to re-scrape the website? (y/n): ").strip().lower()
        
        if choice == 'y':
            # Scrape website
            scraper = WebsiteScraper(WEBSITE_URL, max_pages=MAX_PAGES_TO_SCRAPE)
            documents = scraper.scrape_website()
            
            if documents:
                rag.process_and_store_documents(documents)
            else:
                print("No documents scraped. Exiting...")
                return
        else:
            # Load existing vectorstore
            try:
                rag.load_existing_vectorstore()
            except Exception as e:
                print(f"Error loading vectorstore: {e}")
                print("Please re-scrape the website.")
                return
    else:
        print("No existing FAISS index found. Scraping website...")
        # Scrape website
        scraper = WebsiteScraper(WEBSITE_URL, max_pages=MAX_PAGES_TO_SCRAPE)
        documents = scraper.scrape_website()
        
        if documents:
            rag.process_and_store_documents(documents)
        else:
            print("No documents scraped. Exiting...")
            return
    
    print("\n" + "=" * 50)
    print("Chatbot ready! Ask questions about Agiteks.")
    print("Type 'quit', 'exit', or 'bye' to end\n")
    
    # Initialize conversation history
    conversation_history = []
    
    while True:
        user_input = input("You: ").strip()
        
        if user_input.lower() in ['quit', 'exit', 'bye']:
            print("Chatbot: Goodbye! Have a great day!")
            break
        
        if not user_input:
            continue
        
        try:
            # Retrieve relevant context
            context = rag.retrieve_context(user_input, k=3)
            
            # Create prompt with context
            system_prompt = f"""You are a helpful AI assistant for Agiteks. 
Use the following context from the Agiteks website to answer the user's question.
If the answer is not in the context, say so and provide a general helpful response.

Context from Agiteks website:
{context}

Be concise, friendly, and accurate."""
            
            # Build messages
            messages = [SystemMessage(content=system_prompt)]
            messages.extend(conversation_history[-6:])  # Keep last 3 exchanges
            messages.append(HumanMessage(content=user_input))
            
            # Get response
            response = llm.invoke(messages)
            
            # Update conversation history
            conversation_history.append(HumanMessage(content=user_input))
            conversation_history.append(AIMessage(content=response.content))
            
            print(f"Chatbot: {response.content}\n")
            
        except Exception as e:
            print(f"Error: {e}\n")

if __name__ == "__main__":
    # Check environment variables
    if not os.getenv("GROQ_API_KEY"):
        print("Error: GROQ_API_KEY not found!")
        api_key = input("Enter your Groq API key: ").strip()
        if api_key:
            os.environ["GROQ_API_KEY"] = api_key
        else:
            exit(1)
    
    # Run chatbot
    chatbot_with_rag()